SalesforceCommerceCloud/pwa-kit
# @W-16536103@ [App Extensibility ⚙️] - App-Setup API

# Description

This PR realizes the spike work done [here](https://salesforce.quip.com/1pYFAFLLHbeh) and POC'd [here](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/1972). In short we are adding the the PWA-Kit Application Extensibility API by mirroring the work done to programatically set up the server/express-js application on load.

NOTE: windows tests are failing as the result of previous work for the server extensibility.

# Changes

In this PR we do the following:

1. Create a sample app extension to prove that the react app initialization works, this can also be used as the "template" for the generator when app extension developers want to create a new app extension.
2. Updated the webpack configuration to use a new loader that handles the Application Extension loading.
3. Update both the main (client) and react-rendering (server) files to initialize the application and the configured app extensions.
4. Update the typescript minimal to use this sample extensions so we know it's working. (not required but nice). 

# Types of Changes




# Testing Changes

The best way to see this work in action is to install the `@salesforce/extension-sample` into the `template-typescript-minimal`, you can do this by following these steps:

- Install `@salesforce/extension-sample` into `template-typescript-minimal` by updating the templates package.json file to have this entry under the dev dependencies: `"@salesforce/extension-sample": "1.0.0-dev",`
- Run `npm install` at the root of the mono-repo.
- Next we need to link the new page added in the sample extension, do this by updating `home.tsx` to have the below html:
```
<br />
<br />
Client-side JS works if this counter increments: {counter}
+ <br />
+ <br />
+ App Extensions work if you can navigate to this link:{' '}
+ <Link to="/sample-page">sample page</Link>.
```
- You will also have to import the Link component at the top of the file `import {Link} from 'react-router-dom'`
- Then all you have to do is run the template.

## General
